fix: allow returning of tuples from contract funcs

benesjan opened the pull request:
Returning tuples from contracts didn't work because we forgot to handle it when adding the return value to the hasher.

Fixes #14132

benesjan committed:
fix: allow of returning of tuples from contract funcs

benesjan commented:
* **#14553** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14553?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/14553?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan committed:
works

nventuro commented on noir-projects/noir-contracts/contracts/test/returning_tuple_contract/src/main.nr:
```diff
@@ -0,0 +1,47 @@
+
+    #[private]
+    #[view]
+    fn fn_that_returns_2() -> pub (Field, u32) {
```
```suggestion
    fn fn_that_returns_2() -> (Field, u32) {
```

This is unnecessary

nventuro commented on noir-projects/noir-contracts/contracts/test/returning_tuple_contract/src/main.nr:
```diff
@@ -0,0 +1,47 @@
+
+    #[private]
+    #[view]
+    fn fn_that_returns_1() -> pub (bool) {
+        (true)
```
```suggestion
    fn fn_that_returns_1() -> pub (bool,) {
        (true,)
```

These are not tuples, they are just values sorrounded by parenthesis. You need to do `(x,)` to create a single element tuple.

nventuro replied:
e.g. this works

![image](https://github.com/user-attachments/assets/c2d43167-af12-4ecf-8e31-1501e1b6c6a2)

but this doesn't

![image](https://github.com/user-attachments/assets/04a8a66b-c423-4dd6-b82b-714a533398f5)

nventuro commented on noir-projects/noir-contracts/contracts/test/returning_tuple_contract/src/main.nr:
```diff
@@ -0,0 +1,47 @@
+use aztec::macros::aztec;
+
+// Tests that Aztec.nr handles returning tuples correctly.
+#[aztec]
+pub contract ReturningTuple {
```
I really like the thoroughness of creating a test contract with different variants, but it'd be good to test that this is also handled correctly by the rest of the stack, e.g. txe and e2e tests that call these fns. Not sure where that'd live, do we have tests for handling return values elsewhere?

benesjan committed:
Update noir-projects/noir-contracts/contracts/test/returning_tuple_contract/src/main.nr

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>

benesjan committed:
Update noir-projects/noir-contracts/contracts/test/returning_tuple_contract/src/main.nr

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>

benesjan committed:
removing redundant pub keyword

benesjan committed:
testing tuple value gets correctly decoded

benesjan committed:
fix: correctly decoding string return value (#14603)

This pull request addresses an issue where the ABI decoder incorrectly
returns string values as numeric byte arrays (e.g., [120n, 121n, 122n])
instead of the expected string ("xyz"). The fix modifies the
`AbiDecoder` class to properly parse and convert byte-array-based string
representations into their corresponding string literals.

Fixes #14600

nventuro reviewed (approved):
Lovely
